src/algebra/group_action_hom.lean

```diff
@@ -0,0 +1,252 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+import algebra.group_ring_action
+
+/-!
+# Equivariant homomorphisms
+
+## Main definitions
+
+* `mul_action_hom M X Y`, the type of equivariant functions from `X` to `Y`, notated `X →[M] Y`.
+* `distrib_mul_action_hom M A B`, the type of equivariant additive monoid homomorphisms
+  from `A` to `B`, notated `A →+[M] B`.
+* `mul_semiring_action_hom M R S`, the type of equivariant ring homomorphisms
+  from `R` to `S`, notated `R →+*[M] S`.
+-/
+
+universes u v w u₁
+
+variables (M : Type u) [monoid M]
+variables (X : Type v) [mul_action M X]
+variables (Y : Type w) [mul_action M Y]
+variables (Z : Type u₁) [mul_action M Z]
+variables (A : Type v) [add_monoid A] [distrib_mul_action M A]
+variables (A' : Type v) [add_group A'] [distrib_mul_action M A']
+variables (B : Type w) [add_monoid B] [distrib_mul_action M B]
+variables (B' : Type w) [add_group B'] [distrib_mul_action M B']
+variables (C : Type u₁) [add_monoid C] [distrib_mul_action M C]
+variables (R : Type v) [semiring R] [mul_semiring_action M R]
+variables (R' : Type v) [ring R'] [mul_semiring_action M R']
+variables (S : Type w) [semiring S] [mul_semiring_action M S]
+variables (S' : Type w) [ring S'] [mul_semiring_action M S']
+variables (T : Type u₁) [semiring T] [mul_semiring_action M T]
+variables (G : Type u) [group G] (H : set G) [is_subgroup H]
+
+set_option old_structure_cmd true
+
+/-- Equivariant functions. -/
+@[nolint has_inhabited_instance]
+structure mul_action_hom :=
+(to_fun : X → Y)
+(map_smul' : ∀ (m : M) (x : X), to_fun (m • x) = m • to_fun x)
+
+notation β ` →[`:25 α:25 `] `:0 γ:0 := mul_action_hom α β γ
+
+namespace mul_action_hom
+
+instance : has_coe_to_fun (X →[M] Y) :=
+⟨_, λ c, c.to_fun⟩
+
+variables {M X Y}
+
+theorem map_smul (f : X →[M] Y) (m : M) (x : X) : f (m • x) = m • f x :=
+f.map_smul' m x
+
+@[ext] theorem ext : ∀ {f g : X →[M] Y}, (∀ x, f x = g x) → f = g
+| ⟨f, _⟩ ⟨g, _⟩ H := by { congr' 1, ext x, exact H x }
+
+theorem ext_iff {f g : X →[M] Y} : f = g ↔ ∀ x, f x = g x :=
+⟨λ H x, by rw H, ext⟩
+
+variables (M X)
+
+/-- The identity map is equivariant. -/
+protected def id : X →[M] X :=
```

Comment:
I think `X` should be implicit here. In newer versions of Lean, it no longer has to be explicit if you want the coercion to function to work. So you only need the explicit version if you want to talk about `id` without applying it. This will not be so common, and a type ascription will suffice in those cases.
Same applies to the `id` downstairs.

Reply:
what changed for function coercion?

Reply:
It can now use the type of the argument to fill in meta variables in the type of the `f` that is being coerced to a function.

Reply:
done